// pantheon-helpers/lib/pantheon-helpers-design-docs/audit.js
// Generated by IcedCoffeeScript 1.8.0-c
(function() {
  var a, _;

  _ = require('underscore');

  a = {
    views: {
      audit_by_timestamp: {
        map: function(doc) {
          var entry, out, typ, _i, _len, _ref, _results;
          _ref = doc.audit;
          _results = [];
          for (_i = 0, _len = _ref.length; _i < _len; _i++) {
            entry = _ref[_i];
            typ = doc._id.indexOf('team_') === 0 ? 'team' : 'user';
            out = {
              _id: doc._id,
              name: doc.name,
              entry: entry,
              type: typ
            };
            _results.push(emit(entry.dt, out));
          }
          return _results;
        }
      }
    },
    lists: {
      get_values: function(header, req) {
        var out, row, val;
        out = [];
        while ((row = getRow())) {
          val = row.value;
          out.push(val);
        }
        return JSON.stringify(out);
      }
    },
    rewrites: [
      {
        from: "/audit",
        to: "/_list/get_values/audit_by_timestamp",
        query: {}
      }
    ],
    mixin: function(dd) {
      _.extend(dd.views, a.views);
      _.extend(dd.lists, a.lists);
      dd.rewrites = dd.rewrites.concat(a.rewrites);
      return dd;
    }
  };

  module.exports = a;

}).call(this);
